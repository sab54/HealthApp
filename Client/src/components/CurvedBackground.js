//Client/src/components/CurvedBackground.js
/**
 * CurvedBackground.js
 * 
 * This file defines the `CurvedBackground` component, which creates an animated 
 * curved background effect for the UI. The component uses an animated gradient 
 * background, a lavender glow in the middle, and a white core to create a smooth, 
 * layered visual effect. It also supports translation animation based on the `translateY` prop.
 * 
 * Features:
 * - Displays an animated gradient background using `LinearGradient` from `expo-linear-gradient`.
 * - Includes a lavender-colored glow layer to create a soft, glowing effect.
 * - Features a white core for a clean and smooth finish.
 * - Supports vertical translation animation using the `translateY` prop for dynamic effects.
 * 
 * Props:
 * - `translateY`: The animated translation value (used to move the background vertically).
 * 
 * Dependencies:
 * - `react-native`
 * - `expo-linear-gradient`
 * 
 * Author: Sunidhi Abhange
 */

import React from 'react';
import { View, StyleSheet, Animated } from 'react-native';
import { LinearGradient } from 'expo-linear-gradient';

const CurvedBackground = ({ translateY }) => (
  <Animated.View
    style={[
      styles.container,
      { transform: [{ translateY }] },
    ]}
  >
    {/* Outer gradient */}
    <LinearGradient
      colors={['#e6e6fa', '#ffffff']}
      style={styles.gradient}
      start={{ x: 0.5, y: 0.5 }}
      end={{ x: 0.5, y: 1 }}
    />

    {/* Middle lavender glow */}
    <View style={styles.lavenderGlow} />

    {/* Inner white core */}
    <View style={styles.whiteCore} />
  </Animated.View>
);

export default CurvedBackground;

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 80,
    alignSelf: 'center',
    width: 400,
    height: 400,
  },
  gradient: {
    ...StyleSheet.absoluteFillObject,
    borderRadius: 200,
  },
  lavenderGlow: {
    position: 'absolute',
    top: 40,
    left: 40,
    width: 320,
    height: 320,
    borderRadius: 160,
    backgroundColor: '#e6e6fa',
    opacity: 0.5,
  },
  whiteCore: {
    position: 'absolute',
    top: 100,
    left: 100,
    width: 200,
    height: 200,
    borderRadius: 100,
    backgroundColor: '#ffffff',
    opacity: 0.8,
  },
});
