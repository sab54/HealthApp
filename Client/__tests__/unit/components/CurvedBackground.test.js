/**
 * Client/src/components/__tests__/CurvedBackground.test.js
 *
 * What This Test File Covers:
 *
 * 1) Basic Rendering
 *    - Renders container and the mocked LinearGradient.
 *
 * 2) Transform (translateY)
 *    - Applies translateY value to the Animated.View container.
 *
 * 3) Layers Presence & Key Styles
 *    - Renders lavender glow and white core with expected dimensions/position.
 *
 * 4) Snapshot
 *    - Captures a stable snapshot of the rendered tree.
 */

import React from 'react';
import { render } from '@testing-library/react-native';
import CurvedBackground from 'src/components/CurvedBackground';

// Mock expo-linear-gradient to avoid native issues and to allow prop inspection
jest.mock('expo-linear-gradient', () => ({
  LinearGradient: ({ children, ...props }) => {
    const React = require('react');
    const { View } = require('react-native');
    return React.createElement(View, { ...props, testID: 'linear-gradient' }, children);
  },
}));

// Helper: flatten RN style arrays / objects
const flatten = (style) =>
  Array.isArray(style) ? Object.assign({}, ...style) : style || {};

// Traverse to collect all nodes from toJSON()
const collectNodes = (json) => {
  const out = [];
  const walk = (n) => {
    if (!n || typeof n !== 'object') return;
    out.push(n);
    if (n.children) n.children.forEach(walk);
  };
  walk(json);
  return out;
};

describe('CurvedBackground', () => {
  it('renders container and LinearGradient', () => {
    const { toJSON, getByTestId } = render(<CurvedBackground translateY={0} />);
    const json = toJSON();
    expect(json).toBeTruthy();

    // Mocked gradient is a View with testID
    const gradient = getByTestId('linear-gradient');
    expect(gradient).toBeTruthy();

    // Ensure gradient props carry expected config
    expect(gradient.props.colors).toEqual(['#e6e6fa', '#ffffff']);
    expect(gradient.props.start).toEqual({ x: 0.5, y: 0.5 });
    expect(gradient.props.end).toEqual({ x: 0.5, y: 1 });
  });

  it('applies translateY transform on the container', () => {
    const translateY = 30;
    const { toJSON } = render(<CurvedBackground translateY={translateY} />);
    const json = toJSON();
    const nodes = collectNodes(json);

    // Root container is the first View node in the tree
    const container = nodes.find((n) => n.type === 'View');
    const style = flatten(container.props.style);

    // transform is an array like [{ translateY: <value> }]
    expect(Array.isArray(style.transform)).toBe(true);
    const firstTransform = style.transform[0] || {};
    expect(firstTransform.translateY).toBe(translateY);
  });

  it('renders lavender glow and white core with expected sizes and positions', () => {
    const { toJSON } = render(<CurvedBackground translateY={0} />);
    const nodes = collectNodes(toJSON());

    // Find candidates by style heuristics
    const views = nodes.filter((n) => n.type === 'View');

    // Lavender glow: width 320, height 320, top 40, left 40, opacity 0.5, backgroundColor '#e6e6fa'
    const lavender = views.find((n) => {
      const s = flatten(n.props.style);
      return (
        s.width === 320 &&
        s.height === 320 &&
        s.top === 40 &&
        s.left === 40 &&
        s.opacity === 0.5 &&
        s.backgroundColor === '#e6e6fa'
      );
    });
    expect(lavender).toBeTruthy();

    // White core: width 200, height 200, top 100, left 100, opacity 0.8, backgroundColor '#ffffff'
    const whiteCore = views.find((n) => {
      const s = flatten(n.props.style);
      return (
        s.width === 200 &&
        s.height === 200 &&
        s.top === 100 &&
        s.left === 100 &&
        s.opacity === 0.8 &&
        s.backgroundColor === '#ffffff'
      );
    });
    expect(whiteCore).toBeTruthy();
  });

  it('matches snapshot', () => {
    const tree = render(<CurvedBackground translateY={10} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
