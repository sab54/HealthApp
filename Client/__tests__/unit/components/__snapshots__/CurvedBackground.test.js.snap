// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurvedBackground matches snapshot 1`] = `
<View
  collapsable={false}
  style={
    [
      {
        "alignSelf": "center",
        "height": 400,
        "position": "absolute",
        "top": 80,
        "width": 400,
      },
      {
        "transform": [
          {
            "translateY": 10,
          },
        ],
      },
    ]
  }
>
  <View
    colors={
      [
        "#e6e6fa",
        "#ffffff",
      ]
    }
    end={
      {
        "x": 0.5,
        "y": 1,
      }
    }
    start={
      {
        "x": 0.5,
        "y": 0.5,
      }
    }
    style={
      {
        "borderRadius": 200,
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="linear-gradient"
  />
  <View
    style={
      {
        "backgroundColor": "#e6e6fa",
        "borderRadius": 160,
        "height": 320,
        "left": 40,
        "opacity": 0.5,
        "position": "absolute",
        "top": 40,
        "width": 320,
      }
    }
  />
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "borderRadius": 100,
        "height": 200,
        "left": 100,
        "opacity": 0.8,
        "position": "absolute",
        "top": 100,
        "width": 200,
      }
    }
  />
</View>
`;
